File: 606 - Construct String from Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public String tree2str(TreeNode t) {
        if (t == null) {
            return "";
        }
        String left = "(" + tree2str(t.left) + ")";
        String right = "(" + tree2str(t.right) + ")";
        if (left.equals("()") && right.equals("()")) {
            left = right = "";
        }
        if (right.equals("()")) {
            right = "";
        }
        return t.val + left + right;
    }
}

File: 606 - Construct String from Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    res = []
    def tree2str(self, t: TreeNode) -> str:
        if t is not None:
            self.res.append(str(t.val))
            left = '(' + self.tree2str(t.left) + ')'
            right = '(' + self.tree2str(t.right) + ')'
            if right == '()' and left == '()':
                left = right = ''
            if right == '()':
                right = ''
            return str(t.val) + left + right
        else:
            return ''
            
            